components/object_store/src/disk_cache.rs

```diff
@@ -357,18 +357,36 @@ impl DiskCache {
     }
 
     async fn persist_bytes(&self, filename: &str, payload: Bytes) -> Result<()> {
-        let file_path = std::path::Path::new(&self.root_dir)
-            .join(filename)
+        // When write payload to file, the cache lock is released, so when one thread is
+        // reading, another thread may update it, so we write to tmp file first,
+        // then rename to expected filename to avoid other threads see partial
+        // content.
+        let tmp_filename = format!("{filename}.tmp");
+        let tmp_filepath = std::path::Path::new(&self.root_dir)
+            .join(&tmp_filename)
             .into_os_string()
             .into_string()
             .unwrap();
```

Comment:
Shall we introduce a specific implementation for such `tempfile-for-renaming`? At least, we should ensure the unused temp file to be cleaned if any error occurs after it is created.